fix: Allow CA bundle path without config map

fabiendupont opened the pull request:
**What this PR does / why we need it**:

In the previous code, the AWS_CA_BUNDLE environment variable is only considered if the CA_BUNDLE_CONFIGMAP_NAME environment variable is set, while the doc doesn't explicitely tie they together. This change allows using AWS_CA_BUNDLE independently to leverage a CA bundle certificate that would be preexisting in the image or injected by Kubernetes in a known path.

One use case is leveraging OpenShift service serving certificates, which injects the service CA in the containers via a projected ConfigMap. The service CA is always available in
/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt.

The documentation already says that this variable can be used to set the path to the CA bundle, while the ConfigMap uses a static file name (cabundle.crt) located in the ConfigMap volume mount point.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

Jooho commented on python/kserve/kserve/storage/storage.py:
```diff
@@ -211,19 +211,25 @@ def _download_s3(uri, temp_dir: str) -> str:
-                    ca_bundle_full_path = (
-                        global_ca_bundle_volume_mount_path + "/cabundle.crt"
-                    )
+            isvc_aws_ca_bundle_path = os.getenv("AWS_CA_BUNDLE")
```
The AWS_CA_BUNDLE environment variable is used to specify the path to the CA file, under the assumption that the CA is already present inside the pod. However, shouldn’t there be a logic to check whether this file actually exists?

Also, in the OpenShift use case, it's said that the service-ca.crt file is always present, but in reality, this is controlled by configuration. For example, in this [link](https://github.com/kserve/kserve/blob/master/config/configmap/inferenceservice.yaml#L721), if the value is set to false, the file is not attached to the pod. Our team discussed whether we really can remove this file from runtime. @VedantMahabaleshwarkar can you please share the result ?

fabiendupont replied:
There is a check for the file existence lower. I am just splitting the logic between `AWS_CA_BUNDLE` and `CA_BUNDLE_VOLUME_MOUNT_POINT` environment variables. Then, the `if os.path.exists(ca_bundle_full_path):` block is checking the file.

Jooho replied:
Yes I missed the part! then I am ok with this change :)

Jooho reviewed (dismissed):
/lgtm

Jooho commented:
/ok-to-test

sivanantha321 commented:
@fabiendupont Can you add a unit test for this ?

fabiendupont commented:
> @fabiendupont Can you add a unit test for this ?

@sivanantha321, I didn't find existing tests for S3, so it would mean writing the whole test coverage for S3 and not just for this change. Sounds like something that would deserve its own pull request, no?

fabiendupont committed:
fix: Allow CA bundle path without config map

In the previous code, the AWS_CA_BUNDLE environment variable is only
considered if the CA_BUNDLE_CONFIGMAP_NAME environment variable is
set, while the doc doesn't explicitely tie they together. This change
allows using AWS_CA_BUNDLE independently to leverage a CA bundle
certificate that would be preexistent in the image or injected by
Kubernetes in a known path.

One use case is leveraging OpenShift service serving certificates, which
injects the service CA in the containers via a projected ConfigMap. The
service CA is always available in
/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt.

Signed-off-by: Fabien Dupont <fdupont@redhat.com>

fabiendupont committed:
Add comprehensive tests for AWS_CA_BUNDLE independent usage

- Add 7 test cases covering AWS_CA_BUNDLE functionality without CA_BUNDLE_CONFIGMAP_NAME
- Test AWS_CA_BUNDLE used independently (main PR feature)
- Test backward compatibility with existing ConfigMap functionality
- Test precedence logic when both AWS_CA_BUNDLE and ConfigMap are set
- Test error handling for missing CA bundle files
- Test SSL verification bypass logic
- Test fallback behavior from empty AWS_CA_BUNDLE to ConfigMap
- Tests added to both storage modules for complete coverage
- All tests pass flake8 linting requirements

These tests validate the PR changes that allow AWS_CA_BUNDLE to be used
without requiring CA_BUNDLE_CONFIGMAP_NAME to be set.

Signed-off-by: Fabien Dupont <fdupont@redhat.com>

fabiendupont commented:
@sivanantha321, @Jooho, I added tests and the CI is green.

Jooho commented on python/kserve/kserve/storage/test/test_s3_storage.py:
With [this PR](https://github.com/kserve/kserve/pull/4391), the storage was segregated. 
So I don't think we need this anymore in kserve folder. @spolti am I right?

fabiendupont replied:
But as long as there are tests in kserve/kserve/storage, shouldn't we keep them in sync? I would expect the whole kserve/kserve/storage to be removed at once when all the code is using only kserve/storage, no?

Jooho replied:
right so I don't see any issues to have them now in both places :+1:

Jooho commented on python/kserve/kserve/storage/storage.py:
```diff
@@ -211,19 +211,26 @@ def _download_s3(uri, temp_dir: str) -> str:
+                )
+                ca_bundle_full_path = (
+                    global_ca_bundle_volume_mount_path + "/cabundle.crt"
+                )
```
```suggestion
                ca_bundle_full_path = os.path.join(global_ca_bundle_volume_mount_path, "cabundle.crt")
```

would be better to use `os.path`

fabiendupont replied:
Thanks for the suggestion. I used it while rebasing.

Jooho reviewed (commented):
Thanks for your contribution again!.

@spolti is currently on holiday and unable to review my comment at the moment. Since having test cases in two places isn’t a major issue, I don’t consider this a blocker. If @spolti thinks this part is unnecessary, we can clean it up later.

I just left nit comments :) 
/lgtm

Jooho reviewed (dismissed):
/lgtm

Jooho commented:
/approve

sivanantha321 commented:
Thanks for the contribution !
/lgtm
